src/libcore/array.rs

```diff
@@ -92,6 +96,276 @@ where
+        // into the "alive" region of the array. Reading this element means
+        // that `data[idx]` is regarded as dead now (i.e. do not touch). As
+        // `idx` was the start of the alive-zone, the alive zone is now
+        // `data[alive]` again, restoring all invariants.
+        let out = unsafe { self.data.get_unchecked(idx).read() };
+
+        Some(out)
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let len = self.len_remaining();
+        (len, Some(len))
+    }
+
+    fn count(self) -> usize {
+        self.len_remaining()
+    }
+
+    fn last(mut self) -> Option<Self::Item> {
+        self.next_back()
+    }
+}
+
+#[stable(feature = "array_value_iter_impls", since = "1.38.0")]
+impl<T, const N: usize> DoubleEndedIterator for IntoIter<T, {N}>
+where
+    [T; N]: LengthAtMost32,
+{
+    fn next_back(&mut self) -> Option<Self::Item> {
+        if self.alive.start == self.alive.end {
+            return None;
+        }
+
+        // Decrease end index.
+        //
+        // From the check above we know that `alive.start != alive.end`.
+        // Combine this with the invariant `alive.start <= alive.end`, we know
+        // that `alive.start < alive.end`. As `alive.start` cannot be negative,
+        // `alive.end` is at least 1, meaning that we can safely decrement it
+        // by one. This also maintains the invariant `alive.start <=
+        // alive.end`. However, due to this change, for a short time, the alive
+        // zone is not `data[alive]` anymore, but `data[alive.start..alive.end
+        // + 1]`.
+        self.alive.end -= 1;
+
+        // Read the element from the array. This is safe: `alive.end` is an
+        // index into the "alive" region of the array. Compare the previous
+        // comment that states that the alive region is
+        // `data[alive.start..alive.end + 1]`. Reading this element means that
+        // `data[alive.end]` is regarded as dead now (i.e. do not touch). As
+        // `alive.end` was the end of the alive-zone, the alive zone is now
+        // `data[alive]` again, restoring all invariants.
+        let out = unsafe { self.data.get_unchecked(self.alive.end).read() };
+
+        Some(out)
+    }
+}
+
+#[stable(feature = "array_value_iter_impls", since = "1.38.0")]
+impl<T, const N: usize> Drop for IntoIter<T, {N}>
+where
+    [T; N]: LengthAtMost32,
+{
+    fn drop(&mut self) {
+        // Here we drop all alive elements (but nothing else). It is safe to
+        // call `as_mut_ptr` as all elements within the alive range are
+        // correctly initialized.
+        //
+        // We increment `alive.start` before dropping for exception safety. If
+        // the `drop` panics, we rather want to treat that element as already
+        // droped (with the risk of memory leakage) instead of treating it as
+        // undropped (with the risk of memory unsafety, which is far worse).
+        unsafe {
```

Comment:
Because the implementation here is just a loop, consider using `self.by_ref().for_each(drop);` or similar to do it, as that should be perf-equivalent but avoids needing safety justification.